### DISCUSSION TITLE
Question: is it compatible with the `threading.Lock` ?

### DISCUSSION THREAD

**Discussion by jcao-ai** [x42005e1f/aiologic#4] [Q&A] [answered]
Thanks for this project, according to the doc, I assume that it can be used in multiple event-loops, but can we also use it in a naive python threads like:
```python
from aiologic import Lock

lock = Lock()

with lock:
    # Do some thing
```

**Comment by x42005e1f** [comment, marked as answer]
Yes, of course. You can freely use any high-level synchronization primitives from the `aiologic` module with naive Python threads. Most of them will be more efficient than their analogs, even on CPython, in cases where synchronization is not really needed: those cases where only one thread is actually accessing.

Interface-wise, the difference will be that `acquire()` is called `green_acquire()`, to clearly distinguish between two different multitasking worlds (but the wait methods have different: `condition.wait()` and `await condition`). "Green" because it will work with greenlets when it detects an eventlet/gevent hub in the current thread. But by default it works with threads, so you do not have to worry.

However, if you are looking for an alternative to `threading.Lock`, `aiologic.Lock` will not be suitable for all cases where the original can be used, since it is an alternative to `anyio.Lock`: it will raise a `RuntimeError` if you try to release a lock from a thread or task other than the one you used to acquire the lock. Here is a bit about what primitives can be used as a lock:

1. `aiologic.BinarySemaphore` is an alternative to `threading.Lock` that omits checks for redundant `*_release()` method calls. The fastest synchronization primitive, which is good for cases where you only work with the lock via the context manager or otherwise guarantee the `*_release()` call. You can call `*_release()` from any thread or task.
2. `aiologic.BoundedBinarySemaphore` is like `aiologic.BinarySemaphore`, but with a check that the lock is acquired when the `*_release()` method is called.
3. `aiologic.Lock` is an alternative to `anyio.Lock`: it will check that you call the `*_release()` method from the same thread and task. Like `threading.RLock`, but non-reentrant.
4. `aiologic.RLock` (aka "reentrant lock") is an alternative to `threading.RLock`: you can reacquire it from the same thread and task as many times as you want.

Note that aiologic primitives cannot compete with C-level primitives in terms of speed, so you can expect a slight performance degradation on CPython for Lock / RLock. However, the situation is radically different on PyPy, where `aiologic.BinarySemaphore` can perform faster than `threading.Lock`!
